File: spec/factories/game_results.rb
FactoryBot.define do
  factory :game_result do
    user
    difficulty { "standard" }
    game_mode { "random" }
    score { 1000 }
    time_ms { nil }
    accuracy { 95 }
    wpm { 60 }
    max_combo { 12 }
    correct_cards { 10 }

    trait :random_result do
      game_mode { "random" }
      score { 1000 }
      time_ms { nil }
    end

    trait :timeattack_result do
      game_mode { "timeattack" }
      score { nil }
      time_ms { 30_000 }
    end
  end
end

File: app/frontend/lib/api/game-results.ts
import { postJson } from '@/lib/http';

/**
 * プレイ記録（1プレイ分の最終スコア）を保存する。
 * ログインユーザーがゲームを自然完了したときに呼ぶ（ADR 0005 / #20）。
 * ランキング登録（saveScore）とは独立。timeattack は経過時間を 'time'（ms）で送る。
 */
export interface GameResultPayload {
	game_mode: 'random' | 'timeattack';
	difficulty: 'beginner' | 'standard' | 'advanced';
	score?: number; // random 用
	time?: number; // timeattack 用（ms）
	accuracy: number; // 0..100 の整数パーセント
	wpm: number;
	max_combo: number;
	correct_cards: number;
}

export async function saveGameResult(
	payload: GameResultPayload
): Promise<{ success: boolean; error?: string }> {
	try {
		await postJson('/api/game_results', payload);
		return { success: true };
	} catch (error) {
		console.error('Error saving game result:', error);
		return {
			success: false,
			error: error instanceof Error ? error.message : 'プレイ記録の保存に失敗しました'
		};
	}
}
